src/Model/AbstractClient.php

```diff
@@ -0,0 +1,187 @@
+<?php
+
+declare(strict_types=1);
+
+namespace League\Bundle\OAuth2ServerBundle\Model;
+
+/**
+ * @psalm-consistent-constructor
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ */
+abstract class AbstractClient
+{
+    /**
+     * @var string
+     */
+    private $name;
+
+    /**
+     * @var string
+     */
+    private $identifier;
+
+    /**
+     * @var string|null
+     */
+    private $secret;
+
+    /**
+     * @var list<RedirectUri>
+     */
+    private $redirectUris = [];
+
+    /**
+     * @var list<Grant>
+     */
+    private $grants = [];
+
+    /**
+     * @var list<Scope>
+     */
+    private $scopes = [];
+
+    /**
+     * @var bool
+     */
+    private $active = true;
+
+    /**
+     * @var bool
+     */
+    private $allowPlainTextPkce = false;
+
+    /**
+     * @final
```

Comment:
there's no need for `@final` here, as i mentioned, people would need to extend this, `@final` makes it not possible ( statically speaking, even tho it has 0 effect at runtime )